data/blog/2025/2025-07-23-benefits-of-middleware-over-mvc.md

```diff
@@ -0,0 +1,132 @@
+The **optional view rendering** also helps execution time for middleware, while MVC loads e.g. controller factories, view helpers, translation services and plugins even if you only need a JSON response.
+
+## Enhanced Control Over Request Handling
+
+The **flow** of your application (pipeline) can be much **easier to edit** without rewriting a bunch of code.
+You can **inspect, modify or reject requests** at various points in the pipeline.
+This is particularly handy when processing a request that is dependent on logging, authentication and authorization.
+
+## Middleware Chains
+
+Middleware chains are additional benefits resulting from middleware modularity and enhanced control.
+They can **process a request in a predefined sequence**.
+This is useful for **implementing complex processing logic**, like data processing, logging and authentication in a clean and manageable way.
+
+## Asynchronous Processing
+
+Middleware architectures can help you handle asynchronous processing, which is nowadays increasingly important for **building high-performance web applications**.
+A good example for this scenario is when you need to manage many concurrent connections, while allowing other parts of the application to run in the meantime.
+The downside is that this adds complexity and overhead in the code logic and error handling, making it harder to debug and test, but it can provide benefits in specific cases.
+
+## Better Suited for API-First Development
+
+**MVC is tightly coupled** to the web/MVC context which **difficult to reuse or test** your logic in CLI scripts, workers or other applications.
+**Middleware** can often be a **better fit for API-first development**, where the main interaction with the application is through APIs rather than traditional web pages.
+This is the recommended approach for handling different types of requests, such as RESTful APIs, GraphQL and WebSocket connections.
+
+## Framework Lock-in
+
+MVC has multiple **components tied to the framework**:
+
+- Request & response classes.
+- Routing system.
+- Controllers or service containers.
+- Lifecycle.
+
+Conversely, **middleware is reusable** across currently active frameworks with little or no changes thanks to:
+
+- **Implementing standards** like PSR-7 and PSR-11 which also helps **integrate with future PHP frameworks** that implement them.
+- **Decoupling of handlers** which means they **can be used by API, CLI, worker processes or unit tests** without modification and while not being dependent on HTTP requests, routes, frameworks or controllers.
+- **View rendering is optional or external**, since middleware cares about HTTP requests and responses, not about how they are rendered.
+- The **middleware pipeline**, like the one based on PSR-15, **improves flexibility, customizability and testability** for execution flows.
```

Comment:
PSR-15 does not define a middleware pipeline; that's an implementation detail for frameworks that consume PSR-15.  Maybe "like the ones provided in Laminas Diactoros or Slim Framework"